qubes/storage/reflink.py

```diff
@@ -372,25 +400,38 @@ def _replace_file(dst):
 _remove_file = functools.partial(
     qubes.utils.remove_file, log_level=logging.INFO)
 
-def _make_dir(path):
-    ''' mkdir path, ignoring FileExistsError; return whether we
-        created it.
-    '''
-    with suppress(FileExistsError):
+def _hardlink_file(src, dst):
+    dst_dir = os.path.dirname(dst)
+    _create_dir(dst_dir)
+    os.link(src, dst)
+    qubes.utils.fsync_path(dst_dir)
+    LOGGER.info('Hardlinked file: %r -> %r', src, dst)
+
+def _create_dir(path):
+    try:
+        created = False
         os.mkdir(path)
+        created = True
+    except FileExistsError:
+        if not os.path.isdir(path):
+            raise
+    if created:
         qubes.utils.fsync_path(os.path.dirname(path))
         LOGGER.info('Created directory: %r', path)
-        return True
-    return False
+    return created
```

Comment:
Would it be better to use `mkdirat()` instead?  Not sure if Python supports it.  In any case this can be done in a separate PR.

Reply:
Python would [support it](https://docs.python.org/3/library/os.html#dir-fd), however none of the relink functions attempt to cater to a user doing wild stuff like replacing a directory inside a pool *in the middle of a function call* that's operating on that directory.

> In any case this can be done in a separate PR.

Hmm I generally try to avoid submitting a very small refactor (that modifies the same file) as a separate PR, if the marginal cost of reviewing one more commit in a series would be dwarfed by the overhead (silly merge conflicts, context switching, CI) of another PR.

Reply:
> Python would [support it](https://docs.python.org/3/library/os.html#dir-fd), however none of the relink functions attempt to cater to a user doing wild stuff like replacing a directory inside a pool _in the middle of a function call_ that's operating on that directory.

I would still prefer to use `mkdirat()`, but that might be me being overly pedantic.  Theoretically `mkdirat()` should be a teeny bit faster but it almost certainly does not matter.  That said, `mkdirat()` might even lead to cleaner code.

> > In any case this can be done in a separate PR.
> 
> Hmm I generally try to avoid submitting a very small change (that modifies the same file) as a separate PR, if the marginal cost of reviewing one more commit in a series would be dwarfed by the overhead (silly merge conflicts, context switching, CI) of another PR.

:+1:

Reply:
> `mkdirat()` might even lead to cleaner code.

Nah it gets really convoluted in Python because you have to switch to lower level code patterns, e.g. look at the side-by-side view of https://github.com/rustybird/qubes-core-admin/commit/dir_fd_example

Reply:
Good point, although utility functions would make that easier.